### PR TITLE
In dashboard show api eager load collection with articles

### PR DESCRIPTION
**Pull request by vishaldeepak**
## What type of PR is this? (check all applicable)
- [*] Bug Fix
- [*] Optimization

## Description
Resolves issue #14429 
The N+1 problem is [triggered here](https://github.com/forem/forem/blob/628202646b93c9a25a8142540aa47705e049a55e/app/views/dashboards/_dashboard_article_row.html.erb#L25)

## QA Instructions, Screenshots, Recordings

You need to make sure that you are running server in your local. Make sure you are logged in. You need to have a posts that are connected to series. Load your [dashboard page](http://localhost:3000/dashboard) . You should see a bullet warning down at the bottom left. 

Using this new branch you should not see such a warning. 

### UI accessibility concerns?
None

## Added/updated tests?

- [*] No, and this is why: too small of a change. Let me know otherwise.

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://developers.forem.com) and/or
      [Admin Guide](https://admin.forem.com/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

### CODE REVIEW

**Comment by djuber** [comment]
@vishaldeepak thanks for this! It looks like the change (which follows the bullet warning recommendation) introuced other failures (in the systems tests), for example https://gist.github.com/djuber/98ef24222bdf838a1015875012ae7ccf  

```
     Bullet::Notification::UnoptimizedQueryError:
       user: djuber
       GET /dashboard
       AVOID eager loading detected
         Article => [:collection]
         Remove from your query: .includes([:collection])
       Call stack
         /home/djuber/src/forem/app/lib/middlewares/set_time_zone.rb:10:in `call'
```

It's weird that it's pointing to "set_time_zone" as the culprit... however - that's just the top of the stack where it's calling `@app.call(env)` and is misleading.

The test that's failing is visiting "/dashboard" with various sort parameters - all 6 tests fail (with the error that the mapped selectors aren't seen on the page) - this does not show for me in development mode - I'm looking into what or why this is different.

A [full backtrace](https://gist.github.com/djuber/1f28e3cefa1fc41137d47cb5d046654a) can be seen by adding a breakpoint inside the test - it's almost all library code (not the app code) so it's unclear to me why `@app.call(env)` is seen as the location of the issue

**Review by djuber — CHANGES_REQUESTED** [review]
I think you want to add the following change to config/environments/test.rb in order to get the tests passing

```
1 file changed, 1 insertion(+)
config/environments/test.rb | 1 +

modified   config/environments/test.rb
@@ -79,6 +79,7 @@
     Bullet.add_safelist(type: :n_plus_one_query, class_name: "ActsAsTaggableOn::Tagging", association: :tag)
     # Supress incorrect warnings from Bullet due to included columns: https://github.com/flyerhzm/bullet/issues/147
     Bullet.add_safelist(type: :unused_eager_loading, class_name: "Article", association: :top_comments)
+    Bullet.add_safelist(type: :unused_eager_loading, class_name: "Article", association: :collection)
     Bullet.add_safelist(type: :unused_eager_loading, class_name: "Comment", association: :user)
     # @citizen428: We have not yet resolved all user - profile preloads related to profilge generalization
     Bullet.add_safelist(type: :n_plus_one_query, class_name: "User", association: :profile)
```

This worked for me locally with the failing dashboard user sorts articles system test.

**Comment by vishaldeepak** [comment]
@djuber
This can also be bypassed if we add this to the test case line 5 in file `spec/system/dashboards/user_sorts_dashboard_articles_spec.rb`. Basically I added a collection to one of the articles 
```
  let(:article1) do
    create(:article, user_id: user.id, published_at: 10.minutes.ago, created_at: 1.day.ago, public_reactions_count: 5,
                     page_views_count: 0, comments_count: 100, with_collection: create(:collection))
  end
```
What do you think?

**Comment by djuber** [comment]
> This can also be bypassed if we add this to the test case line 5 in file `spec/system/dashboards/user_sorts_dashboard_articles_spec.rb`. Basically I added a collection to one of the articles

That's not a bad solution - though it leaves open this potential headache for future tests, which probably aren't strictly about the eager loading.

For example - the cypress tests for pinning articles visits the dashboard (so it would need the seed values changed as well)

  Pin an article - Admin User
       should allow to pin another post after the current pinned post is deleted

 There are failing request specs as well (separate from the system spec) and other system specs that incidentally visit the dashboard (and see the error raised) - it's probably both fastest now, and safest for the future, to permit this eager load.

I think these are all the tests that failed (it's possible there are other cypress tests that failed - the test suite did not complete after hitting that error), it seems like a noisy change to add a collection to each of these:

```
rspec ./spec/system/link_for_tags_in_posts_in_notifications_spec.rb:30 # Link on tags for post in notifications when logged in user shows articles
rspec ./spec/system/articles/user_discussion_locks_spec.rb:14 # User discussion locks when an Article does not have a discussion lock shows the discussion lock button in manage
rspec ./spec/system/articles/user_discussion_locks_spec.rb:37 # User discussion locks when an Article has a discussion lock shows the unlock discussion lock button in manage
rspec ./spec/system/dashboards/user_sorts_dashboard_articles_spec.rb:41 # Sorting Dashboard Articles shows articles sorted by default in created_at DESC
rspec ./spec/system/dashboards/user_sorts_dashboard_articles_spec.rb:45 # Sorting Dashboard Articles shows articles sorted by created_at ASC
rspec ./spec/system/dashboards/user_sorts_dashboard_articles_spec.rb:49 # Sorting Dashboard Articles shows articles sorted by comments_count DESC
rspec ./spec/system/dashboards/user_sorts_dashboard_articles_spec.rb:53 # Sorting Dashboard Articles shows articles sorted by page_views_count ASC
rspec ./spec/system/dashboards/user_sorts_dashboard_articles_spec.rb:57 # Sorting Dashboard Articles shows articles sorted by public_reactions_count ASC
rspec ./spec/system/dashboards/user_sorts_dashboard_articles_spec.rb:61 # Sorting Dashboard Articles shows articles sorted by published_at DESC
rspec ./spec/system/articles/user_deletes_an_article_spec.rb:13 # Deleting Article author of article deletes own article
rspec ./spec/requests/dashboard_spec.rb:25 # Dashboards GET /dashboard when logged in renders user's articles
rspec ./spec/requests/dashboard_spec.rb:30 # Dashboards GET /dashboard when logged in shows "STATS" for articles
rspec ./spec/requests/dashboard_spec.rb:38 # Dashboards GET /dashboard when logged in renders the delete button for drafts
rspec ./spec/requests/dashboard_spec.rb:44 # Dashboards GET /dashboard when logged in renders subscriptions for articles with subscriptions
rspec ./spec/requests/dashboard_spec.rb:59 # Dashboards GET /dashboard when logged in renders pagination if minimum amount of posts
rspec ./spec/requests/dashboard_spec.rb:65 # Dashboards GET /dashboard when logged in does not render pagination if less than one full page
rspec ./spec/requests/dashboard_spec.rb:71 # Dashboards GET /dashboard when logged in renders a link to analytics dashboard
rspec ./spec/requests/dashboard_spec.rb:77 # Dashboards GET /dashboard when logged in renders a link to analytics for the org
rspec ./spec/requests/dashboard_spec.rb:85 # Dashboards GET /dashboard when logged in does not render a link to upload a video when enable_video_upload is false
rspec ./spec/requests/dashboard_spec.rb:92 # Dashboards GET /dashboard when logged in does not render a link to upload a video for a recent user
rspec ./spec/requests/dashboard_spec.rb:101 # Dashboards GET /dashboard when logged in as a super admin renders the specified user's articles
rspec ./spec/requests/dashboard_spec.rb:136 # Dashboards GET /dashboard/organization when logged in renders user's organization articles
rspec ./spec/requests/dashboard_spec.rb:144 # Dashboards GET /dashboard/organization when logged in does not render the delete button for other org member's drafts
```

**Comment by vishaldeepak** [comment]
the second option looks more appropriate, i'll check them both out a update the PR.

**Comment by vishaldeepak** [comment]
Looks like the second method did'nt solve the problem for a few test cases.  Thats because the article is unpublished. And hence does not go through this part of code
```
 <% if article.collection %>
          <strong class="fw-medium pl-2">Series:</strong> <%= article.series %>
<% end %>
```   
Here is the build 
https://app.travis-ci.com/github/forem/forem/jobs/536115860
https://app.travis-ci.com/github/forem/forem/jobs/536115860

Think it would be best to `add_whitelist` in test and dev.

**Comment by vishaldeepak** [comment]
Made changes accordingly , let me know if anything else is required.

File: config/environments/development.rb [hunk 1]
```diff
@@ -128,6 +128,7 @@
     Bullet.add_safelist(type: :n_plus_one_query, class_name: "ActsAsTaggableOn::Tagging", association: :tag)
     # Supress incorrect warnings from Bullet due to included columns: https://github.com/flyerhzm/bullet/issues/147
     Bullet.add_safelist(type: :unused_eager_loading, class_name: "Article", association: :top_comments)
+    Bullet.add_safelist(type: :unused_eager_loading, class_name: "Article", association: :collection)
```
Comment (djuber) [inline config/environments/development.rb:131]: Thanks for adding it to the development environment as well!

**Review by djuber — APPROVED** [review]

**Review by juliannatetreault — APPROVED** [review]
Thanks so much for making this change, @vishaldeepak!

**Review by jgaskins — APPROVED** [review]
@vishaldeepak Looks great! Thanks!

**Comment by aitchiss** [comment]
Nice work with this PR @vishaldeepak! There's just one final step before we can merge this, and that's to [sign the contributor agreement above](https://github.com/forem/forem/pull/14677#issuecomment-914152858).

Looking forward to seeing this in the codebase! 😄

**Comment by vishaldeepak** [comment]
Done, signed agreement,thank you

**Pull request merged**